### PR TITLE
minimal pointer capture for mouse events (don't merge now)

### PR DESCRIPTION
**Pull request by robx**
This implements `{set,release}PointerCapture`, compare https://www.w3.org/TR/pointerevents/#widl-Element-setPointerCapture-void-long-pointerId. It doesn't make sense to merge at this point because Elm doesn't even use the methods, compare https://github.com/mpizenberg/elm-pointer-events/issues/4.

This is not a complete or necessarily a correct implementation of these methods, but enough to handle capturing `pointerup` events outside of the element that got the mouse-induced `pointerdown` event. A complete implementation would likely have to do some kind of reverse thing, where touch events are "uncaptured".

### CODE REVIEW

File: elm-pep.js [hunk 1]
```diff
@@ -25,7 +31,13 @@ function addMouseToPointerListener(target, mouseType, pointerType) {
     let pointerEvent = new MouseEvent(pointerType, mouseEvent);
     pointerEvent.pointerId = 1;
     pointerEvent.isPrimary = true;
-    mouseEvent.target.dispatchEvent(pointerEvent);
+
+    let target = mouseEvent.target;
+    if (mouseCaptureTarget !== null) {
+      target = mouseCaptureTarget;
```
Comment (mpizenberg) [inline elm-pep.js]: In that case, I wonder if the `offsetX/Y` properties shouldn't also be adjusted? I cannot test it since chromium and firefox both support `setPointerCapture` from what I've tried. Could you test if safari behave coherently in this regard?

**Comment by mpizenberg** [comment]
Regarding the fact that pointer capture is not supported by elm currently, I don't really mind. The main purpose of this polyfill is to be compatible with elm, but if it can help also pure JavaScript, why not.

**Comment by mpizenberg** [comment]
According to [W3C documentation][w3c-doc], the point 10.3.1 "Implicit Release of Pointer Capture" says that

> "Immediately after firing the pointerup or pointercancel events, a user agent MUST run the steps as if the releasePointerCapture() method has been called with an argument equal to the pointerId property of the pointerup or pointercancel event just dispatched."

So I think we have to reset `mouseCaptureTarget` to null on `pointerup` also.

[w3c-doc]: https://www.w3.org/TR/pointerevents/#dfn-pointer-capture

File: elm-pep.js (hunk 1 above)
Comment (robx) [reply elm-pep.js]: Good point. I'm honestly a bit surprised how `setPointerCapture` seems to be implemented without the pointer events.

Comment (robx) [reply elm-pep.js]: Curiously, the offsetX/Y for mouse-based pointer events seems correct in Safari. I don't understand how/why.

**Comment by robx** [comment]
There's one other thing that's slightly incorrect: It seems you shouldn't be able to capture the pointer on a mouseup event, because the pointer is no longer in the "active" state, but tracking that seems too hard/painful to be worth it.

**Comment by mpizenberg** [comment]
There is something weird bothering me. In the function `definePointerCapture()` there is a test to make sure that `setPointerCapture` is not already implemented: `if (window.Element && !Element.prototype.setPointerCapture)`. In firefox, I've run this test manually on a new empty firefox instance in the console, and it returns false. Apparently `setPointerCapture` is already defined.

However, I've put a console log inside the `if` and when I load `elm-pep.js` in firefox, it does console log. This puzzles me alot.

So I've tried to remove the `definePointerCapture()`. Then the attribute trick `event.target.setPointerCapture(event.pointerId)` is raising an error.

I've tried to simplify a bit the code. Let me know if there are things that you'd prefer to change.
Basically, I've moved the call to `definePointerCapture()` inside the Mouse `if` since I feel it only matters for pointer emulation from mouse.
**EDIT**: it was obviously an error since then the attribute `onpointerdown` triggers error for touch events. It is back to Pointer level.

For the code of definePointerCapture, following the same "mouse" logic, I replaced your manual definition with the already defined functions `{set,release}Capture`. Let me know if you are ok with that. I couldn't find the compatibility table of these ... For me everything seems to work on Firefox (and Chromium obviously) (**EDIT** desktop AND mobile ^^). I'd be glad to know that it works also in Safari :)

**Comment by robx** [comment]
As far as I can tell this seems to work. I've tested Mac desktop Safari, Firefox, Chrome, and Safari on iPhone.

I didn't know about `{set,release}Capture`. Seems non-standard but also only missing in Chrome which has `{set,release}PointerCapture` out of the box. So, perfect!

I probably took that confusing bit of detection logic from the jquery polyfill. Don't really have an idea what's going on there.

Right re `definePointerCapture` on the higher level, I made exactly the same steps. :)

In summary, looking good!
